File: mobile/modules/on-device-face-analysis/ios/OnDeviceFaceAnalysisModule.swift
import ExpoModulesCore
import MLKitFaceDetection
import MLKitVision
import UIKit

public class OnDeviceFaceAnalysisModule: Module {
  public func definition() -> ModuleDefinition {
    Name("OnDeviceFaceAnalysis")

    AsyncFunction("analyzeImageAsync") { (uri: String, promise: Promise) in
      guard let url = URL(string: uri), let image = UIImage(contentsOfFile: url.path) else {
        promise.reject("ERR_FACE_IMAGE", "The captured image could not be decoded.")
        return
      }

      let options = FaceDetectorOptions()
      options.performanceMode = .fast
      options.landmarkMode = .all
      options.classificationMode = .all
      options.minFaceSize = 0.2
      let detector = FaceDetector.faceDetector(options: options)
      let visionImage = VisionImage(image: image)
      visionImage.orientation = image.imageOrientation
      let brightness = self.averageBrightness(image)

      detector.process(visionImage) { faces, error in
        if let error {
          promise.reject("ERR_FACE_ANALYSIS", error.localizedDescription)
          return
        }
        let detected = faces ?? []
        let first = detected.first
        var facePayload: [String: Any]? = nil
        if let face = first {
          facePayload = [
            "x": face.frame.minX / image.size.width,
            "y": face.frame.minY / image.size.height,
            "width": face.frame.width / image.size.width,
            "height": face.frame.height / image.size.height,
            "headEulerAngleX": face.hasHeadEulerAngleX ? face.headEulerAngleX : 0,
            "headEulerAngleY": face.hasHeadEulerAngleY ? face.headEulerAngleY : 0,
            "headEulerAngleZ": face.hasHeadEulerAngleZ ? face.headEulerAngleZ : 0,
            "leftEyeOpenProbability": face.hasLeftEyeOpenProbability ? face.leftEyeOpenProbability : -1,
            "rightEyeOpenProbability": face.hasRightEyeOpenProbability ? face.rightEyeOpenProbability : -1,
            "smilingProbability": face.hasSmilingProbability ? face.smilingProbability : -1
          ]
        }
        promise.resolve([
          "faceCount": detected.count,
          "brightness": brightness,
          "isWellLit": brightness >= 55 && brightness <= 235,
          "imageWidth": image.size.width,
          "imageHeight": image.size.height,
          "face": facePayload as Any
        ])
      }
    }
  }

  private func averageBrightness(_ image: UIImage) -> CGFloat {
    guard let ciImage = CIImage(image: image) else { return 0 }
    let extent = ciImage.extent
    let filter = CIFilter(name: "CIAreaAverage", parameters: [
      kCIInputImageKey: ciImage,
      kCIInputExtentKey: CIVector(cgRect: extent)
    ])
    guard let output = filter?.outputImage else { return 0 }
    var pixel = [UInt8](repeating: 0, count: 4)
    CIContext(options: [.workingColorSpace: NSNull()]).render(
      output,
      toBitmap: &pixel,
      rowBytes: 4,
      bounds: CGRect(x: 0, y: 0, width: 1, height: 1),
      format: .RGBA8,
      colorSpace: nil
    )
    return (0.2126 * CGFloat(pixel[0])) + (0.7152 * CGFloat(pixel[1])) + (0.0722 * CGFloat(pixel[2]))
  }
}

File: mobile/modules/on-device-face-analysis/android/src/main/java/expo/modules/ondevicefaceanalysis/OnDeviceFaceAnalysisModule.kt
package expo.modules.ondevicefaceanalysis

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class OnDeviceFaceAnalysisModule : Module() {
  private val detector by lazy {
    val options = FaceDetectorOptions.Builder()
      .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
      .setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_ALL)
      .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_ALL)
      .setMinFaceSize(0.2f)
      .build()
    FaceDetection.getClient(options)
  }

  override fun definition() = ModuleDefinition {
    Name("OnDeviceFaceAnalysis")

    AsyncFunction("analyzeImageAsync") { uri: String, promise: Promise ->
      val context = appContext.reactContext
      if (context == null) {
        promise.reject("ERR_FACE_CONTEXT", "The native face analyser is not available.", null)
        return@AsyncFunction
      }

      try {
        val imageUri = Uri.parse(uri)
        val image = InputImage.fromFilePath(context, imageUri)
        val bitmap = context.contentResolver.openInputStream(imageUri).use { stream ->
          BitmapFactory.decodeStream(stream)
        } ?: throw IllegalArgumentException("The captured image could not be decoded.")
        val brightness = averageBrightness(bitmap)
        bitmap.recycle()

        detector.process(image)
          .addOnSuccessListener { faces ->
            val face = faces.firstOrNull()
            val facePayload = face?.let {
              mapOf(
                "x" to it.boundingBox.left.toDouble() / image.width,
                "y" to it.boundingBox.top.toDouble() / image.height,
                "width" to it.boundingBox.width().toDouble() / image.width,
                "height" to it.boundingBox.height().toDouble() / image.height,
                "headEulerAngleX" to it.headEulerAngleX.toDouble(),
                "headEulerAngleY" to it.headEulerAngleY.toDouble(),
                "headEulerAngleZ" to it.headEulerAngleZ.toDouble(),
                "leftEyeOpenProbability" to (it.leftEyeOpenProbability.takeIf { value -> value >= 0f }?.toDouble() ?: -1.0),
                "rightEyeOpenProbability" to (it.rightEyeOpenProbability.takeIf { value -> value >= 0f }?.toDouble() ?: -1.0),
                "smilingProbability" to (it.smilingProbability.takeIf { value -> value >= 0f }?.toDouble() ?: -1.0)
              )
            }
            promise.resolve(
              mapOf(
                "faceCount" to faces.size,
                "brightness" to brightness,
                "isWellLit" to (brightness in 55.0..235.0),
                "imageWidth" to image.width,
                "imageHeight" to image.height,
                "face" to facePayload
              )
            )
          }
          .addOnFailureListener { error ->
            promise.reject("ERR_FACE_ANALYSIS", error.message ?: "Face analysis failed.", error)
          }
      } catch (error: Exception) {
        promise.reject("ERR_FACE_IMAGE", error.message ?: "The captured image could not be analysed.", error)
      }
    }
  }

  private fun averageBrightness(bitmap: Bitmap): Double {
    val stepX = maxOf(1, bitmap.width / 64)
    val stepY = maxOf(1, bitmap.height / 64)
    var luminance = 0.0
    var samples = 0
    var y = 0
    while (y < bitmap.height) {
      var x = 0
      while (x < bitmap.width) {
        val pixel = bitmap.getPixel(x, y)
        val red = (pixel shr 16) and 0xff
        val green = (pixel shr 8) and 0xff
        val blue = pixel and 0xff
        luminance += (0.2126 * red) + (0.7152 * green) + (0.0722 * blue)
        samples += 1
        x += stepX
      }
      y += stepY
    }
    return if (samples == 0) 0.0 else luminance / samples
  }
}
